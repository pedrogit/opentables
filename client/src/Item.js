import React from "react";
const Template = require('./common/template');

function Item({template, schema, item}) {
  var newTemplate = new Template(template, schema);

  return (
    <div>
      {newTemplate.replaceValues(item)}
    <br /><br />
    </div>
  );
}

export default Item;
