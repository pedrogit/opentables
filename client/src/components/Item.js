import React from "react";
//import React, {memo} from "react";
import JsxParser from "react-jsx-parser";
import Stack from "@mui/material/Stack";
import Box from "@mui/material/Box";
import { useTheme } from "@mui/material/styles";
import IconButton from '@mui/material/IconButton';
import HighlightOffIcon from '@mui/icons-material/HighlightOff';

import * as Components from "./components";
const Globals = require("../common/globals");

// Define all components in the current scope
// so they can be used directly, without
// namespace (e.g. Components.Text), in the template
for (let name in Components) {
  global[name] = Components[name];
}

function Item({ 
  template, 
  item, 
  rowNb, 
  setLoginState, 
  handleListAuth, 
  handleDeleteItem,
  setViewId
}) {
  const [newItem, setItem] = React.useState(item);
  const [showButtons, setShowButtons] = React.useState(false);

  const theme = useTheme();

  var defaultSx = {
    bgcolor: rowNb % 2 ? "inherit" : "#EEE", 
    padding: 1,
    position: 'relative'
  };

  var buttonSx = {
    position: 'absolute',
    top: '1px',
    right: '1px'
  }

  var handlePatch = function (val, callback) {
    setLoginState({
      open: false,
      msg: {
        severity: "warning",
        title: "Permission denied",
        text:
        'You do not have permissions to edit "' +
        Object.keys(val)[0] +
        '". Please login with valid credentials...',
      },
      action: {
        method: "patch",
        url: "http://localhost:3001/api/opentables/" + newItem[Globals.itemIdFieldName],
        data: val,
        callback: (success, data) => {
          if (success) {
            setItem(data);
            callback(success, data[Object.keys(val)[0]]);
          }
        }
      },
      tryFirst: true
    });
    return false;
  };

  var handleItemAuth = function(action = 'patch', propName, callback) {
    handleListAuth(action = 'patch', item, propName, callback);
  }

  var setBindings = function (item) {
    //console.log('Item setBindings()...');

    // add property name and rest handlers
    var result = {};
    for (var key in item) {
      if (item.hasOwnProperty(key)) {
        result[key] = {
          handlePatch: handlePatch,
          handleItemAuth: handleItemAuth,
          setViewId: setViewId,
          prop: key,
          val: item[key] ? item[key] : "",
        };
      }
    }
    //result.key = result._id.val;
    return result;
  };



  //console.log('Render Item (' +  item.name + ')...');

  return (
    <Box 
      id={"item_" + (newItem ? newItem[Globals.itemIdFieldName] : '')}
      sx={defaultSx}
      onMouseEnter={() => setShowButtons(true)}
      onMouseLeave={() => setShowButtons(false)}
    >
      <JsxParser
        bindings={setBindings(newItem)}
        components={{ ...Components.allComponentsAsJson(), Box, Stack }}
        jsx={template}
        renderInWrapper={false}
        onError={() => {}}
        renderError={({ error }) => <span>{error}</span>}
        showWarnings={false}
        disableKeyGeneration={true} // prevent provided jsx to be reconstructed at every render and to keep their states
        allowUnknownElements={false}
        renderUnrecognized={(tagName) => (
          <span style={{ color: theme.palette.primary.main }}>
            Error: unrecognized tag ({tagName})...
          </span>
        )}
      />
      {showButtons && 
        <Stack sx={buttonSx}>
          <IconButton
            id="deleteItemButton"
            aria-label="delete item" 
            color="inherit"
            onClick={() => handleDeleteItem(newItem[Globals.itemIdFieldName])}
          >
            <HighlightOffIcon />
          </IconButton>
        </Stack>
      }
    </Box>
  );
}

//export default memo(Item);
export default Item;
