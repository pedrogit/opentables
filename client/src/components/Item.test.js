import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import Item from './Item';

describe('Item', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Item />, div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(
      <Item />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});