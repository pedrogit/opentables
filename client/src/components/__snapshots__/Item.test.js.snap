// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item has a valid snapshot 1`] = `
<div
  className="MuiBox-root css-1lf26fz"
  id="item_"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
/>
`;
